hazelcast-client/src/test/java/com/hazelcast/client/ClientDisconnectTest.java

```diff
@@ -0,0 +1,121 @@
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class ClientDisconnectTest extends HazelcastTestSupport {
+
+    private final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
+
+    @After
+    public void cleanup() {
+        hazelcastFactory.terminateAll();
+    }
+
+
+    @Test
+    public void testClientOperationCancelled_whenDisconnected() throws Exception {
+        Config config = new Config();
+        config.setProperty(GroupProperty.CLIENT_ENDPOINT_REMOVE_DELAY_SECONDS.getName(), String.valueOf(Integer.MAX_VALUE));
+        HazelcastInstance hazelcastInstance = hazelcastFactory.newHazelcastInstance();
+        final String queueName = "q";
+
+        final HazelcastInstance client = hazelcastFactory.newHazelcastClient();
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                IQueue<Integer> queue = client.getQueue(queueName);
+                try {
+                    queue.take();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        }).start();
+
+        SECONDS.sleep(2);
+
+        client.shutdown();
+
+        final IQueue<Integer> queue = hazelcastInstance.getQueue(queueName);
+        queue.add(1);
+        //dead client should not be able to consume item from queue
+        assertTrueAllTheTime(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertEquals(queue.size(), 1);
+            }
+        }, 3);
+    }
+
+    @Test
+    public void testClientOperationCancelled_whenDisconnected_lock() throws Exception {
+        Config config = new Config();
+        config.setProperty(GroupProperty.CLIENT_ENDPOINT_REMOVE_DELAY_SECONDS.getName(), String.valueOf(Integer.MAX_VALUE));
+        HazelcastInstance hazelcastInstance = hazelcastFactory.newHazelcastInstance();
+        final String name = "m";
+
+        final IMap<Object, Object> map = hazelcastInstance.getMap(name);
+        final String key = "key";
+        map.lock(key);
+
+        final HazelcastInstance client = hazelcastFactory.newHazelcastClient();
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                IMap<Object, Object> clientMap = client.getMap(name);
+                clientMap.lock(key);
+            }
+        }).start();
+
+        SECONDS.sleep(2);
+
+        client.shutdown();
+
+        map.unlock(key);
+        //dead client should not be able to acquire the lock.
+        assertTrueAllTheTime(new AssertTask() {
```

Comment:
what if thread finishes before this point? will it be a meaningful test in that case?

Reply:
If thread finishes that means it acquired the lock.
Test is expecting that it will not be acquired. Test will fail.